python/ray/util/worker_group.py

```diff
@@ -0,0 +1,308 @@
+            actor._BaseWorkerMixin__execute.remote(construct_metadata))
+        return Worker(actor=actor, metadata=actor_metadata)
+
+    def start(self):
+        """Starts all the workers in this worker group."""
+        if self.workers and len(self.workers) > 0:
+            raise RuntimeError("The workers have already been started. "
+                               "Please call `shutdown` first if you want to "
+                               "restart them.")
+
+        logger.debug(f"Starting {self.num_workers} workers.")
+        self.add_workers(self.num_workers)
+        logger.debug(f"{len(self.workers)} workers have successfully started.")
+
+    def shutdown(self, patience_s: float = 5):
+        """Shutdown all the workers in this worker group.
+
+        Args:
+            patience_s (float): Attempt a graceful shutdown
+                of the workers for this many seconds. Fallback to force kill
+                if graceful shutdown is not complete after this time. If
+                this is less than or equal to 0, immediately force kill all
+                workers.
+        """
+        logger.debug(f"Shutting down {len(self.workers)} workers.")
+        if patience_s <= 0:
+            for worker in self.workers:
+                ray.kill(worker.actor)
+        else:
+            done_refs = [
+                w.actor.__ray_terminate__.remote() for w in self.workers
+            ]
+            # Wait for actors to die gracefully.
+            done, not_done = ray.wait(done_refs, timeout=patience_s)
+            if not_done:
+                logger.debug("Graceful termination failed. Falling back to "
+                             "force kill.")
+                # If all actors are not able to die gracefully, then kill them.
+                for worker in self.workers:
+                    ray.kill(worker.actor)
+
+        logger.debug("Shutdown successful.")
+        self.workers = []
+
+    def execute_async(self, func: Callable[..., T], *args,
+                      **kwargs) -> List[ObjectRef]:
+        """Execute ``func`` on each worker and return the futures.
+
+        Args:
+            func (Callable): A function to call on each worker.
+            args, kwargs: Passed directly into func.
+
+        Returns:
+            (List[ObjectRef]) A list of ``ObjectRef`` representing the
+                output of ``func`` from each worker. The order is the same
+                as ``self.workers``.
+
+        """
+        if len(self.workers) <= 0:
+            raise RuntimeError("There are no active workers. This worker "
+                               "group has most likely been shut down. Please"
+                               "create a new WorkerGroup or restart this one.")
+
+        return [
+            w.actor._BaseWorkerMixin__execute.remote(func, *args, **kwargs)
+            for w in self.workers
+        ]
+
+    def execute(self, func: Callable[..., T], *args, **kwargs) -> List[T]:
```

Comment:
does library still need the blocking method if they can just call ray.get on the async return?

Reply:
Definitely not necessary, but I think it's helpful to include here since we have noticed a common need for blocking calls within SGD already, and I presume it'd be the same in other libraries.

Reply:
This will be replaced by the API above.